client/my-sites/site-settings/settings-section-header.tsx

```diff
@@ -1,40 +1,38 @@
 import { Button } from '@automattic/components';
 import { useTranslate } from 'i18n-calypso';
-import PropTypes from 'prop-types';
+import { FC, PropsWithChildren, ReactNode } from 'react';
 import SectionHeader from 'calypso/components/section-header';
 
-const SettingsSectionHeader = ( {
+interface SettingsSectionHeaderProps extends PropsWithChildren {
+	title: string | ReactNode;
+	id?: string;
+	isSaving?: boolean;
+	showButton?: boolean;
+	disabled?: boolean;
+	onButtonClick?: () => void;
+}
+
+const SettingsSectionHeader: FC< SettingsSectionHeaderProps > = ( {
 	children,
-	disabled,
+	title,
 	id,
 	isSaving,
-	onButtonClick,
 	showButton,
-	title,
-	numberFormat,
-	...buttonProps
```

Comment:
This shouldn't be removed. It might break functionality, and it's an important pattern to follow - [read more](https://dio.la/article/the-open-closed-component-part-1). You'll need to add the types for the forwarded props.

Reply:
I would love to simplify it so I had reviewed all the usages of `SettingsSectionHeader` component and I was sure it's not used anywhere.

Reviewed it one more time after reading your comment and noticed one usage:
https://github.com/Automattic/wp-calypso/blob/61b6db614e1c72c33788bdb2482b8277f607f540/client/my-sites/site-settings/form-general.jsx#L593-L600

`data-tip-target` is forwarded to the button.

I'll bring it back if I have to but I don't like the fact that component props are on the same level as button props. WDYT about introducing `buttonProps` prop for button props?

The pattern you mentioned is great where you have similar components, like `Button` and for example `SubmitButton` which extends it. In our case we have `SettingsSectionHeader` that _contains_ a button, not extends it.

Reply:
I strongly prefer sticking to open/closed, otherwise composition (not just with other components, but also with a lot of libraries and frameworks) is more difficult in the long term.

For example, if you ever wanted to add a tooltip to this component (with something like Radix and Ariakit) you then need to create an additional wrapper component that does expose this API and handle the forwarding to that `buttonProps` prop manually by splitting the props appropriately, which is not only tedious to do, but also tedious to type properly.

Reply:
The open/closed component idea, does it mean that absolutely all components should always follow this pattern, and should always support the full DOM interface of one of the children elements? That sounds quite radical.

In this case `buttonProps` were passed not to the top-level element of `SettingsSectionHeader`, but to the inner `Button`. Which is not even mandatory, it appears only when the `showButton` prop is `true`.

Reply:
I brought the open/closed component pattern back but don't like it for this case for many reasons:
1. The `title` prop exists in both components but with different type definition so I needed to `Omit< ButtonProps, 'title' >` to avoid conflict.
2. Now `SettingsSectionHeader` supports two props for click handling `onClick` from `Button` component and `onButtonClick` from `SettingsSectionHeader`, it's confusing.
3. `SettingsSectionHeader` supports all the `Button` props but this component is not actually a button, it's a header with a button inside it. What if we'll add let's say a `Badge` component to it. Does it mean the `SettingsSectionHeader` should handle `Badge` props as well?

Reply:
Thanks for the extra details. I hadn't noticed those. If that's the case, then I don't think it should be open/closed indeed and the buttonProps strategy seems perfectly fine to me (let's just make sure we're not breaking previous usage).

Open/closed is definitely not meant as something that should be applied in every single component, not at all. The main use case exists in component libraries (like the components package) where almost everything should abide by it, and beyond that it's more of a case by case basis.

In this case, since it seemed to be pre-existing (partially, at least the props rest/spread), the suggested approach seemed like a deviation from it, but with the added context I think it is an improvement and I'm all for it 👍